### PR TITLE
Sharded BlueStore - 3 - resharding tool

### PR DESCRIPTION
**Pull request by aclamk**
This is 3/3 of code that allows BlueStore to use RocksDB in sharded mode.
Third part does:
- add 'reshard' command to ceph_bluestore_tool that modifies column families count per prefix

### CODE REVIEW

**Comment by aclamk** [comment]
Part 1: #29084 
Part 2: #29086
Part 3: #29087

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -5526,50 +5526,26 @@ int BlueStore::_open_db(bool create, bool to_repair_db, bool read_only)
...
+    if (create) {
+      if (cct->_conf.get_val<bool>("bluestore_rocksdb_cf")) {
+        std::string sharding_schema = cct->_conf.get_val<string>("bluestore_rocksdb_cfs");
+        r = write_meta("sharding", sharding_schema);
+        if (r < 0)
+          return r;
       }
+    }
+    std::string sharding_schema;
+    r = read_meta("sharding", &sharding_schema);
```
Comment (jdurgin) [inline src/os/bluestore/BlueStore.cc]: is it possible to alter per-cf rocksdb options (other than number of shards) without resharding? e.g. if we wanted to tune a given cf for better performance

File: doc/man/8/ceph-kvstore-tool.rst [hunk 2]
```diff
@@ -15,46 +15,57 @@ Synopsis
...
-leveldb/rocksdb's data (like OSD's omap) offline.
+:program:`ceph-kvstore-tool` is a tool to interact with key-value database used internally by ceph.
+Key-value databases are part of object store instances.
+Modes *leveldb* and *rocksdb* are for FileStore and *bluestore-kv* is for BlueStore.
+FileStore uses DB only for OMAP, while BlueStore also preserves internal metadata in DB.
+
+OSD must be offline to manipulate on its key-value DB.
+Operations are not logically validated. Modifications of BlueStore's metadata are unadvised.
+
+Column Families
```
Comment (jdurgin) [inline doc/man/8/ceph-kvstore-tool.rst]: is there already a command that would let us list the columns?

File: src/common/options.cc [hunk 3]
```diff
@@ -4708,8 +4708,8 @@ std::vector<Option> get_global_options() {
     .set_description("Enable use of rocksdb column families for bluestore metadata"),
 
     Option("bluestore_rocksdb_cfs", Option::TYPE_STR, Option::LEVEL_DEV)
-    .set_default("M= P= L=")
-    .set_description("List of whitespace-separate key/value pairs where key is CF name and value is CF options"),
+    .set_default("O7= M5= L= P= b= C=")
+    .set_description("List of whitespace-separate key/value pairs where key is CF name and value is CF options. When CF name is followed with a number, eg. M5, then multiple CFs are created M-0, M-1, ... M-4. No number after CF name means only single CF is created eg. L-0. Prefixes not listed remain part of 'default' column family."),
```
Comment (liewegas) [inline src/common/options.cc]: use .set_long_description() here for the detailed description of how these values are constructued

File: doc/man/8/ceph-kvstore-tool.rst (hunk 2 above)
Comment (aclamk) [reply doc/man/8/ceph-kvstore-tool.rst]: @jdurgin There was not. Now it is and is called "list-columns"

**Comment by neha-ojha** [comment]
http://pulpito.ceph.com/nojha-2019-09-04_01:35:15-rados-wip-sharding-3-tool-reshard-3-distro-basic-smithi/ - don't see any related failures

doing a rerun of failed/dead: http://pulpito.ceph.com/nojha-2019-09-06_19:07:44-rados-wip-sharding-3-tool-reshard-3-distro-basic-smithi/

**Comment by aclamk** [comment]
[On DNM] Performance testing on incerta07 has led me to believe that there is some serious problem. 
I got SIGSEGVs from tc_newarray from RocksDB compation thread (2 times):
https://gist.github.com/aclamk/fd6a28c1b10434a671202f2084ff969a
And once data corruption.
`tc_newarray` happened even in case when sharding was not enabled.

**Comment by markhpc** [comment]
@aclamk Here are performance and compaction results from Officinalis.  We don't see a huge performance difference as compaction is not the primary bottleneck in either the RGW or RBD tests, but we do see a great reduction in compaction overhead in both cases.

https://docs.google.com/spreadsheets/d/1FYFBxwvE1i28AKoLyqrksHptE1Z523NU3Fag0MELTQo/edit?usp=sharing

**Pull request closed without merging**